### ISSUE TITLE
Will this work on server too ?

### ISSUE THREAD

**Issue by Xample** [nspangler/autoreconnect#1]
I have a server connecting on another server:

``` js
// server code
var backendConnection = DDP.connect('http://localhost:5000');
var newCollection = new Mongo.Collection('collectionName', backendConnection);
...
```

Will this package also work on the server side ?

**Comment by nspangler** [comment]
I'm not really quite too sure what functionality you're interested in. Do you want the Meteor server to try and reconnect to `http://localhost:5000` when the connection is lost?

**Comment by Xample** [comment]
Correct. I am wondering if you package supercharge the DDP.connect function in which case it would reconnect both from the client side and the server (if that one is acting as a client of course).
[EDIT] After watching at the code, I see this should only work on the client side.

**Comment by nspangler** [comment]
You are correct. As of right now this package is just a simple client side `Meteor.reconnect` in a timed loop. However, I am interested in adding this specific server side functionality. Besides connecting to an external mongodb collection, do you see any other use case for such functionality?

**Comment by Xample** [comment]
Well, not really. Also I have a strange but good behaviour. I have a slave and a master server, the slave connecting to the master and reading the master's collection. If I disconnect the master at some point or start it after the slave, the slave one will automatically reconnect and synchronise the subscription. I do not know if this will always work or if there is a 2 minutes timeout. I need to experiment this. A reconnect timer would ensure someone is explicitly telling the slave to reconnect, in anyway this can't be bad, if the slave is connected already nothing will happen.

**Comment by Xample** [comment]
Okay here is the log:
I20150823-15:16:42.464(4)? App started with settings:
I20150823-15:16:42.465(4)? {}
I20150823-15:16:42.466(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:16:42.466(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:16:42.466(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:16:47.934(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:16:56.103(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:17:19.427(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:18:12.647(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:20:10.951(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED
I20150823-15:24:41.968(4)? stream error Network error: ws://localhost:5000/websocket: connect ECONNREFUSED

Therefore per default, the DDP clients tries to reconnect in what looks like to be an exponential time i.e. the delay between each tries double if no server has been detected. I have no idea on the maximum delay between reconnections if there is one.

**Comment by nspangler** [comment]
That is the exact same behavior that a client (browser) exhibits. I'm sure adding the same functionality to the DDP connect on the server will be fairly easy.  What does `DDP.connect('http://localhost:5000')` return? Does it return something like "connected", "disconnected" ?

**Comment by Xample** [comment]
It returns a connection you can use within your collection for example:

``` js
var backendConnection = DDP.connect(server);
var ACollection = new Mongo.Collection('acollection', backendConnection);
```

Here are the docs:

To call methods on another Meteor application or subscribe to its data sets, call DDP.connect with the URL of the application. DDP.connect returns an object which provides:
- subscribe - Subscribe to a record set. See Meteor.subscribe.
- call - Invoke a method. See Meteor.call.
- apply - Invoke a method with an argument array. See Meteor.apply.
- methods - Define client-only stubs for methods defined on the remote server. See Meteor.methods.
- status - Get the current connection status. See Meteor.status.
- reconnect - See Meteor.reconnect.
- disconnect - See Meteor.disconnect.
- onReconnect - Set this to a function to be called as the first step of reconnecting. This function can call methods which will be executed before any other outstanding methods. For example, this can be used to re-establish the appropriate authentication context on the new connection.

By default, clients open a connection to the server from which they're loaded. When you call Meteor.subscribe, Meteor.status, Meteor.call, and Meteor.apply, you are using a connection back to that default server.

**NOW**
Calling reconnect has no effect if the connection is connected, so doing something like:

``` js
var retryIntervalInMinutes = 1
Meteor.setInterval(function(){ backendConnection.reconnect(); }, retryIntervalInMinutes *60*1000);
```

Would theoretically ensure keeping alive the connection at least every minutes (instead of the exponential time).

**Comment by nspangler** [comment]
Unfortunately in your code above, the `setInterval` would be a wasted process when the connection does exist, which hopefully will be the majority of the time. Instead, we need a `Tracker.autorun` on the server which tracks the status of the connection and then on `connection : error` tries over the `setInterval` (as stated in your code above). This should be fairly simple if we use https://github.com/peerlibrary/meteor-server-autorun . Thoughts?

**Comment by Xample** [comment]
My browser crashed while I was answering. In summary

If you want to implement your solution, the Meteor.status is reactive, you can get the status of the connection right after doing a DDP.connect in which the returned object do have it's status. Then make use of the external `Tracker.autorun` package, watch if when it does changed the status is `waiting` and if so, start a timer which will ensure to call `reconnect()` on the connection object returned by `DDP.connect`

I still keep in mind using a watchdog with a given interval, repetitively calling reconnect
- does not relies on any other package (for the server)
- does really not cost a lot
- is simple to handle

I took a look to see if we could hook the reconnect on an event (`onReconnect`) but I am not sure this would be appropriate as most of the events are aimed to be used by the end programmer, overriding the method we might put there.

**Comment by nspangler** [comment]
Awesome. I will have some time this next weekend and will write some of the corresponding code based on your thoughts and research. I will let you know when the new package is pushed.
